Tool cannot sign wildcard paths

For some reason when run within an Azure DevOps release task it is not working, erroring out on saying KeyVault URL not specified.

As you can see from the log it is specified (also tried -kvu) but still erroring on that step:

```
2019-05-02T18:02:54.3830516Z nugetkeyvaultsigntool sign D:\a\r1\a\_Alexa.NET-master\drop\*.nupkg -file-digest sha256 -timestamp-rfc3161 "http://timestamp.digicert.com" --timestamp-digest sha256 --azure-key-vault-url "REMOVED_FOR_THIS_ISSUE_POST" --azure-key-vault-client-id "***" --azure-key-vault-client-secret "***" --azure-key-vault-certificate "***"
2019-05-02T18:03:00.1332781Z fail: NuGetKeyVaultSignTool.Program[0]
2019-05-02T18:03:00.1333613Z       Key Vault URL not specified
```

Weird, that same failure of leading '-' wasn't a problem locally...but I changed it and still failing in the task, but now the log shows me nothing...just runs with
```Waiting for console output from an agent...```
for about 1.5 minutes and then fails:
```
2019-05-02T20:14:50.1505601Z Unhandled Exception: System.AggregateException: One or more errors occurred. (Operation returned an invalid status code 'InternalServerError') ---> Microsoft.Azure.KeyVault.Models.KeyVaultErrorException: Operation returned an invalid status code 'InternalServerError'
2019-05-02T20:14:50.1507204Z    at Microsoft.Azure.KeyVault.KeyVaultClient.GetCertificateWithHttpMessagesAsync(String vaultBaseUrl, String certificateName, String certificateVersion, Dictionary`2 customHeaders, CancellationToken cancellationToken)
2019-05-02T20:14:50.1507285Z    at Microsoft.Azure.KeyVault.KeyVaultClientExtensions.GetCertificateAsync(IKeyVaultClient operations, String vaultBaseUrl, String certificateName, CancellationToken cancellationToken)
2019-05-02T20:14:50.1507397Z    at NuGetKeyVaultSignTool.SignCommand.SignAsync(String packagePath, String outputPath, String timestampUrl, HashAlgorithmName signatureHashAlgorithm, HashAlgorithmName timestampHashAlgorithm, SignatureType signatureType, Boolean overwrite, String keyVaultCertificateName, String keyVaultUrl, String keyVaultClientId, String keyVaultClientSecret, String keyVaultAccessToken) in D:\a\1\s\NuGetKeyVaultSignTool.Core\SignCommand.cs:line 66
2019-05-02T20:14:50.1507500Z    at NuGetKeyVaultSignTool.Program.<>c__DisplayClass0_1.<<Main>b__3>d.MoveNext() in D:\a\1\s\NuGetKeyVaultSignTool\Program.cs:line 91
2019-05-02T20:14:50.1507722Z    --- End of inner exception stack trace ---
2019-05-02T20:14:50.1507784Z    at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
2019-05-02T20:14:50.1507855Z    at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
2019-05-02T20:14:50.1507912Z    at NuGetKeyVaultSignTool.Program.Main(String[] args) in D:\a\1\s\NuGetKeyVaultSignTool\Program.cs:line 155
2019-05-02T20:14:55.5888013Z ##[debug]Exit code: -532462766
2019-05-02T20:14:55.5929338Z ##[debug]Leaving Invoke-VstsTool.
2019-05-02T20:14:55.6456593Z ##[error]Cmd.exe exited with code '-532462766'.
```

--- a/NuGetKeyVaultSignTool.Core/NuGetKeyVaultSignTool.Core.csproj
+++ b/NuGetKeyVaultSignTool.Core/NuGetKeyVaultSignTool.Core.csproj
@@ -9,6 +9,7 @@
     <PackageReference Include="Microsoft.Azure.KeyVault" Version="3.0.3" />
     <PackageReference Include="Microsoft.IdentityModel.Clients.ActiveDirectory" Version="4.5.1" />
     <PackageReference Include="NuGet.Packaging" Version="5.2.0-xprivate.60006" />
+    <PackageReference Include="NuGet.Protocol" Version="5.0.0" />
     <PackageReference Include="Portable.BouncyCastle" Version="1.8.5" />
     <PackageReference Include="RSAKeyVaultProvider" Version="1.1.22" />
   </ItemGroup>

--- a/NuGetKeyVaultSignTool.Core/SignCommand.cs
+++ b/NuGetKeyVaultSignTool.Core/SignCommand.cs
@@ -10,6 +10,7 @@ using Microsoft.Extensions.Logging;
 using NuGet.Common;
 using NuGet.Packaging.Signing;
 using ILogger = Microsoft.Extensions.Logging.ILogger;
+using NuGet.Protocol;
 
 namespace NuGetKeyVaultSignTool
 {
@@ -76,38 +77,42 @@ namespace NuGetKeyVaultSignTool
 
         public async Task<bool> SignAsync(string packagePath, string outputPath, string timestampUrl, HashAlgorithmName signatureHashAlgorithm, HashAlgorithmName timestampHashAlgorithm, bool overwrite, X509Certificate2 publicCertificate, System.Security.Cryptography.RSA rsa)
         {
-            var fileName = Path.GetFileName(packagePath);
-            logger.LogInformation($"{nameof(SignAsync)} [{fileName}]: Begin Signing {packagePath}");
+            var packagesToSign = LocalFolderUtility.ResolvePackageFromPath(packagePath);
+            
             var signatureProvider = new KeyVaultSignatureProvider(rsa, new Rfc3161TimestampProvider(new Uri(timestampUrl)));
 
             var request = new AuthorSignPackageRequest(publicCertificate, signatureHashAlgorithm, timestampHashAlgorithm);
 
             string originalPackageCopyPath = null;
-            try
+            foreach (var package in packagesToSign)
             {
-                originalPackageCopyPath = CopyPackage(packagePath);
-
-                using (var options = SigningOptions.CreateFromFilePaths(originalPackageCopyPath, outputPath, overwrite, signatureProvider, new NuGetLogger(logger, fileName)))
-                {
-                    await SigningUtility.SignAsync(options, request, CancellationToken.None);
-                }
-            }
-            catch (Exception e)
-            {
-                logger.LogError(e, e.Message);
-                return false;
-            }
-            finally
-            {
+                logger.LogInformation($"{nameof(SignAsync)} [{package}]: Begin Signing {Path.GetFileName(package)}");
                 try
                 {
-                    FileUtility.Delete(originalPackageCopyPath);
-                }
-                catch
-                {
-                }
+                    originalPackageCopyPath = CopyPackage(package);
 
-                logger.LogInformation($"{nameof(SignAsync)} [{fileName}]: End Signing {packagePath}");
+                    using (var options = SigningOptions.CreateFromFilePaths(originalPackageCopyPath, package, overwrite, signatureProvider, new NuGetLogger(logger, package)))
+                    {
+                        await SigningUtility.SignAsync(options, request, CancellationToken.None);
+                    }
+                }
+                catch (Exception e)
+                {
+                    logger.LogError(e, e.Message);
+                    return false;
+                }
+                finally
+                {
+                    try
+                    {
+                        FileUtility.Delete(originalPackageCopyPath);
+                    }
+                    catch
+                    {
+                    }
+
+                    logger.LogInformation($"{nameof(SignAsync)} [{package}]: End Signing {Path.GetFileName(package)}");
+                }
             }
 
             return true;

--- a/NuGetKeyVaultSignTool.Core/VerifyCommand.cs
+++ b/NuGetKeyVaultSignTool.Core/VerifyCommand.cs
@@ -13,13 +13,14 @@ using NuGet.Packaging;
 using NuGet.Packaging.Signing;
 using Microsoft.Extensions.Logging;
 using ILogger = Microsoft.Extensions.Logging.ILogger;
+using NuGet.Protocol;
 
 namespace NuGetKeyVaultSignTool
 {
     public class VerifyCommand
     {
         readonly ILogger logger;
-        
+
         public VerifyCommand(ILogger logger)
         {
             this.logger = logger;
@@ -37,44 +38,50 @@ namespace NuGetKeyVaultSignTool
                 throw new ArgumentNullException(nameof(buffer));
             }
 
-            
+
             var trustProviders = new ISignatureVerificationProvider[]
             {
                 new IntegrityVerificationProvider(),
                 new SignatureTrustAndValidityVerificationProvider()
             };
             var verifier = new PackageSignatureVerifier(trustProviders);
+
+            var allPackagesVerified = true;
+
             try
             {
                 var result = 0;
-                using (var package = new PackageArchiveReader(file))
+                var packagesToVerify = LocalFolderUtility.ResolvePackageFromPath(file);
+
+                foreach (var packageFile in packagesToVerify)
                 {
-                    var verificationResult = await verifier.VerifySignaturesAsync(package, SignedPackageVerifierSettings.GetVerifyCommandDefaultPolicy(), CancellationToken.None);
-
-
-                    if (verificationResult.IsValid)
+                    using (var package = new PackageArchiveReader(packageFile))
                     {
-                        return verificationResult.IsValid;
-                    }
-                    else
-                    {
-                        var logMessages = verificationResult.Results.SelectMany(p => p.Issues).Select(p => p .AsRestoreLogMessage()).ToList();
-                        foreach (var msg in logMessages)
+                        var verificationResult = await verifier.VerifySignaturesAsync(package, SignedPackageVerifierSettings.GetVerifyCommandDefaultPolicy(), CancellationToken.None);
+
+                        if (verificationResult.IsValid)
                         {
-                            buffer.AppendLine(msg.Message);
+                            allPackagesVerified = true;
                         }
-                        if (logMessages.Any(m => m.Level >= NuGet.Common.LogLevel.Warning))
+                        else
                         {
-                            var errors = logMessages.Where(m => m.Level == NuGet.Common.LogLevel.Error).Count();
-                            var warnings = logMessages.Where(m => m.Level == NuGet.Common.LogLevel.Warning).Count();
+                            var logMessages = verificationResult.Results.SelectMany(p => p.Issues).Select(p => p.AsRestoreLogMessage()).ToList();
+                            foreach (var msg in logMessages)
+                            {
+                                buffer.AppendLine(msg.Message);
+                            }
+                            if (logMessages.Any(m => m.Level >= NuGet.Common.LogLevel.Warning))
+                            {
+                                var errors = logMessages.Where(m => m.Level == NuGet.Common.LogLevel.Error).Count();
+                                var warnings = logMessages.Where(m => m.Level == NuGet.Common.LogLevel.Warning).Count();
 
-                            buffer.AppendLine($"Finished with {errors} errors and {warnings} warnings.");
+                                buffer.AppendLine($"Finished with {errors} errors and {warnings} warnings.");
 
-                            result = errors;
+                                result = errors;
+                            }
+                            allPackagesVerified = false;
                         }
-                        return false;
                     }
-
                 }
             }
             catch (Exception e)
@@ -82,6 +89,8 @@ namespace NuGetKeyVaultSignTool
                 logger.LogError(e, e.Message);
                 return false;
             }
+
+            return allPackagesVerified;
         }
     }
 }

--- a/NuGetKeyVaultSignTool/Program.cs
+++ b/NuGetKeyVaultSignTool/Program.cs
@@ -122,12 +122,6 @@ namespace NuGetKeyVaultSignTool
                         return -1;
                     }
 
-                    if (!File.Exists(file.Value))
-                    {
-                        application.Error.WriteLine("File does not exist");
-                        return -1;
-                    }
-
                     var cmd = new VerifyCommand(logger);
                     var buffer = new StringBuilder();
                     var result = await cmd.VerifyAsync(file.Value, buffer);

--- a/NuGetKeyVaultSignTool/Program.cs
+++ b/NuGetKeyVaultSignTool/Program.cs
@@ -32,9 +32,9 @@ namespace NuGetKeyVaultSignTool
                 var packagePath = signConfiguration.Argument("packagePath", "Package to sign.");
                 var outputPath = signConfiguration.Option("-o | --output", "The output file. If omitted, overwrites input.", CommandOptionType.SingleValue);
                 var force = signConfiguration.Option("-f | --force", "Overwrites a sigature if it exists.", CommandOptionType.NoValue);
-                var fileDigestAlgorithm = signConfiguration.Option("-fd | --file-digest", "The digest algorithm to hash the file with.", CommandOptionType.SingleValue);
+                var fileDigestAlgorithm = signConfiguration.Option("-fd | --file-digest", "The digest algorithm to hash the file with. Default option is sha256", CommandOptionType.SingleValue);
                 var rfc3161TimeStamp = signConfiguration.Option("-tr | --timestamp-rfc3161", "Specifies the RFC 3161 timestamp server's URL. If this option (or -t) is not specified, the signed file will not be timestamped.", CommandOptionType.SingleValue);
-                var rfc3161Digest = signConfiguration.Option("-td | --timestamp-digest", "Used with the -tr switch to request a digest algorithm used by the RFC 3161 timestamp server.", CommandOptionType.SingleValue);
+                var rfc3161Digest = signConfiguration.Option("-td | --timestamp-digest", "Used with the -tr switch to request a digest algorithm used by the RFC 3161 timestamp server. Default option is sha256", CommandOptionType.SingleValue);
                 var signatureType = signConfiguration.Option("-st | --signature-type", "The signature type (omit for author, default. Only author is supported currently).", CommandOptionType.SingleValue);
                 var azureKeyVaultUrl = signConfiguration.Option("-kvu | --azure-key-vault-url", "The URL to an Azure Key Vault.", CommandOptionType.SingleValue);
                 var azureKeyVaultClientId = signConfiguration.Option("-kvi | --azure-key-vault-client-id", "The Client ID to authenticate to the Azure Key Vault.", CommandOptionType.SingleValue);
@@ -46,7 +46,7 @@ namespace NuGetKeyVaultSignTool
                 {
                     if (string.IsNullOrWhiteSpace(packagePath.Value))
                     {
-                        logger.LogError("All arguments are required");
+                        logger.LogError("Path to file(s) to sign are requried");
                         return -1;
                     }
 

